### DISCUSSION TITLE
RP2040 - GFX lib not compileable

### DISCUSSION THREAD

**Discussion by Deadeye5589** [moononournation/Arduino_GFX#276] [Q&A] [answered]
Hello,

I'm trying to compile a project I have started for the Lilygo T-Display-S3 now in combination with an RP2040 Pi Pico Board and a Waveshare 1.9 Inch ST7789 display. Which is basically identical to the Lilygo integrated display. Since nothing seems to work. I have decided to use a basic approch and tried the samples. Since I'm using LVGL on the Lilygo board my educated guess was to open the LVGLHelloWord example. But neither the Arduino MBED OS, nor the RP2040 board definitions from earlephilhower can be compiled. 

When using the RP2040 definitions the error message is:
_{standard input}: Assembler messages:
{standard input}: Error: unaligned opcodes detected in executable segment_

while with the MBED OS definition I basically get the same message for all the SPI pin declarations:
Arduino_RPiPicoSPI(int8_t dc = GFX_NOT_DEFINED, int8_t cs = GFX_NOT_DEFINED, int8_t sck = PIN_SPI0_SCK, int8_t mosi = PIN_SPI0_MOSI, int8_t miso = PIN_SPI0_MISO, spi_inst_t *spi = spi0); // Constructor

what I changed when swapping to the RP2040 board was the following lines of code in the example:

`#define GFX_BL DF_GFX_BL // default backlight pin, you may replace DF_GFX_BL to actual backlight pin

/* More dev device declaration: https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration */
#if defined(DISPLAY_DEV_KIT)
Arduino_GFX *gfx = create_default_Arduino_GFX();
#else /* !defined(DISPLAY_DEV_KIT) */

/* More data bus class: https://github.com/moononournation/Arduino_GFX/wiki/Data-Bus-Class */
Arduino_DataBus *bus = create_default_Arduino_DataBus();

/* More display class: https://github.com/moononournation/Arduino_GFX/wiki/Display-Class */
Arduino_GFX *gfx = new Arduino_ILI9341(bus, DF_GFX_RST, 0 /* rotation */, false /* IPS */);

#endif /* !defined(DISPLAY_DEV_KIT) */`

to:

`#define GFX_BL 0

Arduino_DataBus *bus = new Arduino_RPiPicoSPI(4 /* DC */, 5 /* CS */, 2 /* SCK */, 3 /* MOSI */, UINT8_MAX /* MISO */, spi0 /* spi */);
Arduino_GFX *gfx = new Arduino_ST7789(bus, 1 /* RST */, 3 /* rotation */, true /* IPS */, 170, 320, 0, 80);
`
Any idea what could be missing since both LVGL and GFX are supposed to support the RP2040 already.

**Comment by moononournation** [comment]
This is an known issue, you need to select optimizations to -O2 to compile success. But I don't know why.

**Reply by Deadeye5589** [reply]
Thank you for the fast reply. That was indeed the right solution. Only for the "earlephilhower" RP2040 library though, but that shall be enough for me. The MBED Core seems to have a general problem with your library, since even the moste basic examples without LVGL background won't compile.

Since the display is running I was able to figure out the rigth offsets. Maybe you want this information as a additional member to the ST7789 LCD display class information in the Wiki?

**1.9" LCD 170x320**

```
Arduino_GFX *gfx = new Arduino_ST7789(bus, 1 /* RST */, 3 /* rotation */, true /* IPS */, 170, 320, 35, 0, 35, 0);
```

**Reply by moononournation** [reply, marked as answer]
For RP2040 compile error, please change the optimization option to "-O2"
